project_milestone/tests/test_project_milestone.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2019 Patrick Wilson <patrickraymondwilson@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests.common import TransactionCase
+
+
+class TestProjectMilestone(TransactionCase):
+
+    def setUp(self):
+        super(TestProjectTaskDependency, self).setUp()
+
+        # Create a new project with the tests
+        test_project = self.env['project.project'].create({
+            'name': 'TestProject'})
+
+        # Add a test Milestone to the project
+        test_project_milestone = self.env['project.milestone'].create({
+            'name': 'TestMilestone',
+            'project_id': test_project.id})
+
+        # Create a test Task to the project with the MILESTONE
+        test_project_task = self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': test_project.id,
+            'milestone_id': test_project_milestone.id})
+
+    def test_01_creation_check(self):
+        # Perform Checks
+        self.assertEqual(test_project.name, 'TestProject')
+        self.assertEqual(test_project_milestone.name, 'TestMilestone')
+        self.assertEqual(test_project_milestone.project_id, test_project.id)
+        self.assertEqual(test_project_task.name, 'TestTask')
+        self.assertEqual(test_project_task.project_id, test_project.id)
+        self.assertEqual(test_project_task.milestone_id,
+                         test_project_milestone.id)

```

Comment:
Some of these tests are not really needed. I see that you are new to unit testing, so let me explain. :wink: 

You are working on top of a framework that already handles the case where calling `model.create({"name": "TestMilestone"})` actually creates a record with that value in the name. The framework already tests that, so you don't need to. It's obvious that it's going to happen.

The things you actually need to test are the special behaviors related to the logic in your module. Some examples:

1. When I create a milestone, does it get assigned the correct sequence?
1. When I  change a task stage, does the milestone get the correct percentage of completion?
1. When I group with a default project in the context, does the grouping return the expected value?

These are some of the points in your module that alter the default behavior, so those are the ones you should test instead.

Reply:
@Yajo your suggestions are great, thank you. I understand what you are saying and i'm working on improving this, but i'm not sure how to do test #3 with the group view. Do you have any examples on how to code that?